src/background-view.cpp

```diff
@@ -0,0 +1,355 @@
+
+#include <wayfire/util/log.hpp>
+
+extern "C"
+{
+#include <wlr/config.h>
+#if WLR_HAS_XWAYLAND
+#define class class_t
+#define static
+#include <wlr/xwayland.h>
+#undef static
+#undef class
+#endif
+}
+
+struct process
+{
+    nonstd::observer_ptr<wf::view_interface_t> view;
+    struct wl_client *client;
+    pid_t pid;
+};
+
+static std::map<wf::output_t *, struct process> procs;
+
+class wayfire_background_view : public wf::plugin_interface_t
+{
+    const std::string transformer_name = "background-view";
+    wf::config::option_base_t::updated_callback_t option_changed;
+    wf::option_wrapper_t<std::string> command{"background-view/command"};
+    wf::option_wrapper_t<std::string> file{"background-view/file"};
+    struct wl_event_source *signal = nullptr;
+
+    public:
+    void init() override
+    {
+        grab_interface->name = transformer_name;
+        grab_interface->capabilities = 0;
+
+        option_changed = [=] ()
+        {
+            if (procs[output].view)
+            {
+                procs[output].view->close();
+                kill(procs[output].pid, SIGINT);
+                procs[output].view = nullptr;
+            }
+
+            procs[output].client = client_launch((std::string(command) + add_arg_if_not_empty(file)).c_str());
+        };
+
+        command.set_callback(option_changed);
+        file.set_callback(option_changed);
+
+        output->connect_signal("map-view", &view_mapped);
+
+        if (!signal)
+        {
+            signal = wl_event_loop_add_signal(wf::get_core().ev_loop, SIGCHLD, sigchld_handler, &procs);
+        }
+
+        procs[output].client = client_launch((std::string(command) + add_arg_if_not_empty(file)).c_str());
+        procs[output].view = nullptr;
+    }
+
+    wf::signal_connection_t view_mapped{[this] (wf::signal_data_t *data)
+    {
+        auto view = get_signaled_view(data);
+#if WLR_HAS_XWAYLAND
+        wlr_surface *wlr_surface = view->get_wlr_surface();
+        bool is_xwayland_surface = wlr_surface_is_xwayland_surface(wlr_surface);
+#endif
+        for (auto& o : wf::get_core().output_layout->get_outputs())
+        {
+            if (procs[o].client == view->get_client()
+#if WLR_HAS_XWAYLAND
+                || (is_xwayland_surface &&
+                /* For this match to work, the client must set _NET_WM_PID */
+                procs[o].pid == wlr_xwayland_surface_from_wlr_surface(wlr_surface)->pid)
+#endif
+                )
+            {
+#if WLR_HAS_XWAYLAND
+                if (is_xwayland_surface)
+                    view->set_decoration(nullptr);
+#endif
+                /* Move to the respective output */
+                wf::get_core().move_view_to_output(view, o);
+
+                /* A client should be used that can be resized to any size
+                 * If we set it fullscreen, the screensaver would be inhibited
+                 * so we send a resize request that is the size of the output
+                 */
```

Comment:
you can set the client to maximized. I think that then most clients will obey the size you give them (even if they don't otherwise, for example weston-terminal)

Reply:
I tried set_tiled(true); before calling set_geometry() and this just makes mpv the wrong size (output.height - panel.height, but centered, so borders at top and bottom). With weston-terminal this works without setting maximized. I don't think it should set clients maximized.